### PR TITLE
Add conditional method renaming rector

### PR DESCRIPTION
**Pull request by markstory**
We have a few renames in CakePHP that are parameter dependent. This rector lets a method be renamed based on the first parameter value. I've only supported string values as that is the current requirements.

### CODE REVIEW

File: packages/CakePHP/src/Rector/MethodCall/RenameMethodCallBasedOnParameterRector.php [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        $this->methodNamesByTypes = $methodNamesByTypes;
+    }
+
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition(
+            'Changes method calls based on matching the first parameter value.',
+            [
+                new CodeSample(
+                    <<<'PHP'
```
Comment (TomasVotruba) [inline packages/CakePHP/src/Rector/MethodCall/RenameMethodCallBasedOnParameterRector.php]: Here `ConfiguredCodeSample` is better

File: packages/CakePHP/tests/Rector/MethodCall/RenameMethodCallBasedOnParameterRector/RenameMethodCallBasedOnParameterRectorTest.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            RenameMethodCallBasedOnParameterRector::class => [
+                '$methodNamesByTypes' => [
+                    SomeModelType::class => [
+                        'getParam' => [
+                            'matchParameter' => 'paging',
+                            'replaceWith' => 'getAttribute',
+                        ],
+                        'withParam' => [
+                            'matchParameter' =>  'paging',
+                            'replaceWith' => 'withAttribute',
```
Comment (TomasVotruba) [inline packages/CakePHP/tests/Rector/MethodCall/RenameMethodCallBasedOnParameterRector/RenameMethodCallBasedOnParameterRectorTest.php]: Keys are better done in underscore, as it's yaml convention to separate from PHP method/property naming

File: packages/CakePHP/src/Rector/MethodCall/RenameMethodCallBasedOnParameterRector.php [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        $argumentType = $config['first_argument_type_to_set'];
+        $argumentValue = $methodCall->args[0]->value;
+
+        if ($argumentType === 'array' && $argumentValue instanceof Array_) {
+            return $config['set'];
+        }
+
+        return $config['get'];
+    }
+}
```
Comment (TomasVotruba) [inline packages/CakePHP/src/Rector/MethodCall/RenameMethodCallBasedOnParameterRector.php]: Could we somehow reuse big chunk of code from `RenameMethodCallRector`? E.g. via service extraction?

File: packages/CakePHP/tests/Rector/MethodCall/RenameMethodCallBasedOnParameterRector/Fixture/fixture.php.inc [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    $config = $object->getAttribute('paging');
+    $object->withAttribute('paging', 'value');
+
+    $config = $object->getParam($value);
+
+    $config = $object->getParam('other');
+    $object->withParam('other', 'value');
+}
+
```
Comment (TomasVotruba) [inline packages/CakePHP/tests/Rector/MethodCall/RenameMethodCallBasedOnParameterRector/Fixture/fixture.php.inc]: A skipped test fixture in own file would be nice

File: packages/CakePHP/src/Rector/MethodCall/RenameMethodCallBasedOnParameterRector.php (hunk 3 above)
Comment (markstory) [reply packages/CakePHP/src/Rector/MethodCall/RenameMethodCallBasedOnParameterRector.php]: I'm dumb this code is totally unused.

**Comment by TomasVotruba** [comment]
Please rebased on `master`, that should fix CI

**Comment by TomasVotruba** [comment]
Looks good to me, thanks :+1:

**Pull request merged**